resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -65,10 +65,10 @@ protected void doWork(ServletOutputStream sos)
       {
          try
          {
+            stream.write(bytes, offset, length);
```

Comment:
Most async 3.1 examples write the bytes after the `ServletOutputStream#isReady` is true.  At least that's what I recall.  Can you elaborate on why this move was done and why it works?

Reply:
I agree with this. It seems odd to write before the output stream is ready.

Reply:
This is the implementation which was there till Resteasy v4.7.3.Final(please check [here](https://github.com/resteasy/resteasy/blob/4.7.3.Final/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L67-L72)) until [this commit](https://github.com/resteasy/resteasy/commit/f87dc2aa3dfe5a25ab2059850b6c38eb30e4231b) which went into 4.7.4.Final. So please understand that this is not something new that I am proposing. The only change that I am having in this PR which was not there in the 4.7.3.Final version is [this check here](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L250-L255) without which we will have this [bug](https://github.com/resteasy/resteasy/pull/2954).

Reply:
My main intention to clarify the above is to convey that these are not my changes and so this is better explained by someone from resteasy team. 

But having said that, I can give an explanation based on my understanding from the debugging experience I have for this issue and [another issue](https://github.com/resteasy/resteasy/pull/2954) we reported earlier on this. 
The doWork for writing the data to an outputstream is only called from [one place which always checks if the outputstream is ready before calling it](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L265-L267). So this logic is not actually writing the bytes before the stream is ready. 
Now coming to why the future completion needs the check whether the output stream is ready, while writing large responses the outputstream is taking sometime to get to a ready state after the write execution. So in the new implementation where [future is completed immediately after write(without checking if the outpustream is ready)](https://github.com/resteasy/resteasy/blob/4.7/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L71-L72), the server completes the response(because of the future completion) before the write is completely done. So having [the check for outputstream being ready](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L101-L105) ensures that the previous write is completely done before the future completes and so we can avoid malformed chunk issues in the response.

Reply:
Having the write after an `isReady` check is fine, is we feel strongly about it. But the key here, is as @dnagarajan89 mentioned, is that a successful `write` call per-se doesn't seem to guarantee completion of write (at-least on the CoyoteOutputStream) and so, for good measure, we would need to check again (with the isReady) post writing.

Reply:
Thank you @dnagarajan89. I see what you're saying now. The write not being complete though seems odd as the write should be synchronous. If it's not we still have a race condition in that the future may not complete if `sos.isReady()` returns `false`. I think we're just getting lucky that it returns `true`.

The main issue I see here is we're trying to use a reactive solution in a non-reactive class. I think we need to look at this in general and possibly re-write some things to improve it.

Reply:
Thanks @dnagarajan89 and @anthochristen .  This diff alone made this hard to see, but onWritePossible javadoc really clears up the original code quite a bit.

Reply:
@jamezp this issue is happening during under some extreme concurrent situation. But as you said the write is synchronus and I don't see a problem with that. The issue here is sometimes the CoyoteOutputStream is taking sometime to empty its buffer(which can be checked with [isReady](https://tomcat.apache.org/tomcat-9.0-doc/api/org/apache/catalina/connector/CoyoteOutputStream.html#isReady())). So when it happens we should not be completing the future as it causes  this exception "org.apache.http.MalformedChunkCodingException: Unexpected content at the end of chunk" on the client side. Thats why as @anthochristen had mentioned earlier, we should be completing the future only by checking isReady in the outputstream. If that is not returning `true` and if the future doesn't gets completed, as per this [contract](https://tomcat.apache.org/tomcat-9.0-doc/api/org/apache/catalina/connector/CoyoteOutputStream.html#isReady()) there will be a callback to onWritePossible which will [call the flushQueue](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L274) which completes the future that is [pending completion in the lastAsyncOperation](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L249).

Reply:
The problem I see is that the `write()` and `isReady()` are implementation specific. If the `CoyoteOutputStream` is taking too long to flush the buffer, then the write is not synchronous or we need to flush synchronously after the write. The fact that `isReady()` returns `true` if you move the `stream.write()` is just luck. The call to `flushQueue()` does not solve this as the `AsyncOperation` has already been removed from the queue so the future associated with it may still never be completed. While there is the `lastAsyncOperation` the value may change if there is more than one op in queue.

One option might be if `isReady()` returns `false` we add a new `AsyncOperation` to the queue which would simply complete the future.

Reply:
"The fact that `isReady()` returns true if you move the `stream.write(`) is just luck." - This statement is not correct. In the failure case([reproducer](https://github.com/dnagarajan89/resteasy-tomcat/tree/RESTEASY-3071)) the `isReady()` after the `write()` is not returning `true` even after moving the `stream.write()` out. The future completion happens [here](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L248-L249), where the `flushQueue()` is called from `onWritePossible()`.

RE your concern of AsyncOperation#future may never get completed, the `lastAsyncOperation` value cannot change before the future get completed because that is the first thing we are doing in [flushQueue()](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L248-L249) before we reassign the `lastAsyncOperation` from the queue [here](https://github.com/resteasy/resteasy/blob/ea829f87d61ef60aa66b207c471cbea95b70278e/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L265).

Reply:
Do you see any other scenario where this lastAsyncOperation value might change where the future is still pending completion.

Reply:
Since the value of `lastAsyncOperation` is assigned in a loop and is synchronized on `this` then it could definitely change before `flushQueue()` is invoked again. There's definitely a race where the future may never complete.

If there are say 5 ops in the queue and the first op goes through, but `isReady()` returns `false` so the future is not complete. The loop would continue and reassign the `lastAsyncOperation`. The listener may be waiting for the lock to release on `flushQueue()`. That leaves the first ops future incomplete.

Reply:
@jamezp did you get a change to look into this? Please let us know your concerns if any.

Reply:
Sorry, my week has been getting away from me. I've only had a chance to take a quick look, but I'll get to it this week.